internal/services/jsonrpc/usersvc/service.go

```diff
@@ -511,14 +567,56 @@ func resultMap(r *sql.ResultSet) []map[string]any {
 	return m
 }
 
+func (svc *Service) verifyCallChallenge(challenge [32]byte) *jsonrpc.Error {
+	svc.challengeMtx.Lock()
+	defer svc.challengeMtx.Unlock()
+
+	challengeTime, ok := svc.challenges[challenge]
+	if !ok {
+		return jsonrpc.NewError(jsonrpc.ErrorCallChallengeNotFound, "invalid challenge", nil)
+	}
+
+	// remove the challenge from the list
+	delete(svc.challenges, challenge)
+
+	// ensure that challenge is not expired
+	if time.Now().After(challengeTime) {
+		return jsonrpc.NewError(jsonrpc.ErrorCallChallengeExpired, "challenge expired", nil)
+	}
+
+	return nil
+}
+
 func (svc *Service) Call(ctx context.Context, req *userjson.CallRequest) (*userjson.CallResponse, *jsonrpc.Error) {
-	body, msg, err := convertActionCall(req)
+	body, msg, err := unmarshalActionCall(req)
 	if err != nil {
 		// NOTE: http api needs to be able to get the error message
 		return nil, jsonrpc.NewError(jsonrpc.ErrorInvalidParams, "failed to convert action call: "+err.Error(), nil)
 
 	}
 
+	// Authenticate by validating the challenge was server-issued, and verify
+	// the signature on the serialized call message that include the challenge.
+	if svc.authCall {
+		// The message must have a sig, sender, and challenge.
+		if msg.Signature == nil || len(msg.Sender) == 0 {
+			return nil, jsonrpc.NewError(jsonrpc.ErrorCallChallengeNotFound, "signed call message with challenge required", nil)
+		}
+		if len(msg.Body.Challenge) != 32 {
+			return nil, jsonrpc.NewError(jsonrpc.ErrorInvalidCallChallenge, "incorrect challenge data length", nil)
+		}
+		// Ensure we issued the message's challenge.
+		if err := svc.verifyCallChallenge([32]byte(msg.Body.Challenge)); err != nil {
+			return nil, err
+		}
+		sigtxt := transactions.CallSigText(body.DBID, body.Action,
+			msg.Body.Payload, msg.Body.Challenge)
+		err = ident.VerifySignature(msg.Sender, []byte(sigtxt), msg.Signature)
+		if err != nil {
+			return nil, jsonrpc.NewError(jsonrpc.ErrorInvalidCallSignature, "invalid signature on call message", nil)
```

Comment:
Since we have `transactions.CallMessage.AuthType` and `transactions.CallMessage.Signature.AuthType`, we need to have some conditional logic here. Either:

- The `AuthType` we pass to the engine is  `transactions.CallMessage.AuthType` in the case of public mode, and  `transactions.CallMessage.Signature.AuthType` in case of private mode.
OR
- We check that  `transactions.CallMessage.Signature.AuthType` and  `transactions.CallMessage.AuthType` are equal if we are in private mode.

Right now, we only pass  `transactions.CallMessage.AuthType` to the engine, which doesn't necessarily match the authtype used in private mode